### PR TITLE
replay: ncurses GUI 

### PR DESCRIPTION
**Pull request by deanlee**
~it's still very drafty.~

update:

replay

### CODE REVIEW

**Comment by adeebshihadeh** [comment]
Wow, looks great so far!

File: selfdrive/ui/replay/replay.cc [hunk 1]
```diff
@@ -269,10 +293,11 @@ void Replay::startStream(const Segment *cur_segment) {
   // write CarParams
   it = std::find_if(events.begin(), events.end(), [](auto e) { return e->which == cereal::Event::Which::CAR_PARAMS; });
   if (it != events.end()) {
+    car_name_ = (*it)->event.getCarParams().getCarName();
```
Comment (adeebshihadeh) [inline selfdrive/ui/replay/replay.cc:296]: use carParams.carFingerprint instead

**Comment by deanlee** [comment]
Show different types of alerts
alert

**Comment by adeebshihadeh** [comment]
Is there a nice library we can use on top or instead of ncurses to make this more readable?

**Comment by deanlee** [comment]
I think the c library is more clear and straightforward than other libraries in this case. there are less than 200 lines of code in consoleui responsible for displaying the interface, the usage pattern of other libraries is very similar to C (set attr, move to position, print, close attr).  they basically don't make the code smaller or easier to read.
I'll cleanup the current code to make it easier to read.

File: selfdrive/ui/replay/consoleui.cc [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+  sm.update(0);
+  if (sm.updated("carState")) {
+    write_item(0, 25, "SPEED:   ", util::string_format("%.2f", sm["carState"].getCarState().getVEgo()), " m/s");
+  }
+  if (sm.updated("liveParameters")) {
+    auto p = sm["liveParameters"].getLiveParameters();
+    write_item(1, 0, "STIFFNESS: ", util::string_format("%.2f", p.getStiffnessFactor() * 100), " deg");
+    write_item(1, 25, "STEER RATIO: ", util::string_format("%.2f", p.getSteerRatio()), "");
+    auto angle_offsets = util::string_format("%.2f|%.2f", p.getAngleOffsetAverageDeg(), p.getAngleOffsetDeg());
+    write_item(2, 0, "ANGLE OFFSET(AVG|INSTANCE): ", angle_offsets, " deg");
```
Comment (robbederks) [inline selfdrive/ui/replay/consoleui.cc:162]:
```suggestion
    write_item(2, 0, "ANGLE OFFSET(AVG|INSTANT): ", angle_offsets, " deg");
```

File: selfdrive/ui/replay/consoleui.cc [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+
+  sm.update(0);
+
+  if (status != Status::Paused) {
+    status = (sm.updated("carState") || sm.updated("liveParameters")) ? Status::Playing : Status::Waiting;
+  }
+  auto [status_str, status_color] = status_text[status];
+  write_item(0, 0, "STATUS:    ", status_str, "      ", false, status_color);
+  write_item(0, 25, "TIME:  ", format_seconds(replay->currentSeconds()),
+             (" / " + format_seconds(replay->totalSeconds())).c_str(), true);
```
Comment (adeebshihadeh) [inline selfdrive/ui/replay/consoleui.cc:158]: can we add current segment too? like 1/11

**Pull request merged**